# Generator
## 1. Introduction
This is also a resolution to handle async prgoramming.
```js
function* generator() {
  yield "Bonjour";
  yield "Madam";
  return "Bonjour!"
}

let gen = generator();
gen.next();
gen.next();
gen.next();
// { value: 'Bonjour', done: false }
// { value: 'Madam', done: false }
// { value: 'Bonjour!', done: true }
```

### 1.1 yeild expression
由于 Generator 函数返回的遍历器对象，只有调用next方法才会遍历下一个内部状态，所以其实提供了一种可以暂停执行的函数。yield表达式就是暂停标志。

- 遍历器对象的next方法的运行逻辑如下。
  - 遇到yield表达式，就暂停执行后面的操作，并将紧跟在yield后面的那个表达式的值，作为返回的对象的value属性值。
  - 下一次调用next方法时，再继续往下执行，直到遇到下一个yield表达式。
  - 如果没有再遇到新的yield表达式，就一直运行到函数结束，直到return语句为止，并将return语句后面的表达式的值，作为返回的对象的value属性值。
  - 如果该函数没有return语句，则返回的对象的value属性值为undefined。

> 需要注意的是，yield表达式后面的表达式，只有当调用next方法、内部指针指向该语句时才会执行，因此等于为 JavaScript 提供了手动的“惰性求值”（Lazy Evaluation）的语法功能。
```js
function* f() {
  console.log("Performed!");
}

// 1） 调用f()并进行赋值时，f()不会执行，因其实Generator函数
let gen = f();

// 2） 只有调用其返回的对象的.next方法时，f()才会执行
setTimeout(()=>{
  gen.next()
},2000)
// Performed!
// { value: undefined, done: true }
```
> 注意： yield表达式只能用在 Generator 函数里面，用在其他地方都会报错。

### 1.2 与Iterator接口的关系
- 任意一个对象的`Symbol.iterator`方法，等于该对象的遍历器生成函数，调用该函数会返回该对象的一个遍历器对象。
- 由于 Generator 函数就是遍历器生成函数，因此可以把 Generator 赋值给对象的Symbol.iterator属性，从而使得该对象具有 Iterator 接口。
```js
const iterator = {};
iterator[Symbol.iterator] = function* () {
  yield 1;
  yield 2;
  yield 3;
  return 'end';
};

[...iterator]
// [ 1, 2, 3 ]
```
- Generator 函数执行后，返回一个遍历器对象。该对象本身也具有Symbol.iterator属性，执行后返回自身。
```js
const gen = function* () {}
const g = gen();
g[Symbol.iterator]() === g;  // true
```

## 2. next method params
yield表达式无返回值，OR 总返回`undefined`;
next可带一个参数，作为上一个yield表达式返回值
> 例一:
```js
const gen = function* () {
  let r1 = yield 1;
  console.log(r1);
  let r2 = yield r1;
};

let g = gen();
g.next();     // 第一次返回值
g.next(2);    // 2作为第一次yield方法的返回值，并传入赋值给r1
g.next();
// {value: 1, done: false}
// {value: 4, done: false}
// {value: undefined, done: true}
```
> 例二：通过next方法参数，注入值
```js
function* f(x) {
  let y = 2 * (yield (x+ 1));
  let z = yield (y / 3);
  return (x + y + z);
} 

let gen = f(5);    // x: 5
gen.next();   // return 6  
gen.next(12); // return 8, yield return 12, y = 24
gen.next(13); // return 42, yield return 15, z= 13
// {value: 6, done: false}
// {value: 8, done: false}
// {value: 42, done: true}
```

> 例三：注入值
```js
function* dataConsumer() {
  console.log("Started!");
  console.log(`1. ${yield}`);
  console.log(`2. ${yield}`);
  return "End!";
}

let gen = dataConsumer();
gen.next();
gen.next("A");
gen.next("B");
/*Print records: */
// "Started!"
// "1. A"
// "2. B"
// "End!"

/* reutrn results: */
// {value: undefined, done: false}
// {value: undefined, done: false}
// {value: "End!", done: true}
```

> 例四：首次调用next就有返回值
```js
function wrapper(generatorFunction) {
  return function (...args) {
    let generatorObject = generatorFunction(...args);
    // 直接调用一次  输入参数给yield返回值用
    generatorObject.next();
    return generatorObject;
  }
}

const warpped = wrapper(function* () {
  console.log(`First input: ${yield}`);
  return "DONE";
})
warpped().next("hello");
```

## 3. for...of循环
`for...of`循环可自动遍历Generator函数运行时生成的`Iterator`对象，且此时不再需要调用`next`方法
```js
function* foo () {
  yield 'Hello';
  yield '你好';
  yield 'Bonjour';
  yield 'Hola';
  return `Let's start!`;
}

for (let w of foo()) {
  console.log(w);
}
// Hello
// 你好
// Bonjour
// Hola
```
> 此时返回四个yield结果，第5个返回值为return结果的done为true，则for...of就会终止

`Generator`函数为对象添加遍历接口
```js
// 方法一：
function* objectEntries(obj) {
  let propKeys = Reflect.ownKeys(obj);
  for (let key of propKeys) {
    yield [key, obj[key]];
  }

  /* This method does not work. */ 
  // propKyes.forEach(key => {
  //   yield [key, obj[key]];
  // })
}
for (let pair of objectEntries({name:"tom", age: 18})) {
  console.log(pair);
}
// [ 'name', 'tom' ]
// [ 'age', 18 ]

// 方法二：
function* objEntries() {
  let keys = Reflect.ownKeys(this);
  for (let key of keys) {
    yield [ key, this[key] ];
  }
}

let tom = { name: "Tom", age: 18 };
tom[Symbol.iterator] = objEntries;
for (let pair of tom) {
  console.log(pair);
}
// [ 'name', 'Tom' ]
// [ 'age', 18 ]
// [ Symbol(Symbol.iterator), [GeneratorFunction: objectEntries] ]
```

`fibonacci`数列实现
```js
function* fibonacci() {
  let [prev, curr] = [0, 1];
  for (;;) {
    yield prev;
    [prev, curr] = [curr, prev + curr];
  }
}
for (let n of fibonacci()) {
  if (n > 10) break;
  console.log(n);
}
// 0 1 1 2 3 5 8
```

除了`for...of`调用遍历器接口，还有扩展运算符`...`, `结构赋值`和`Array.from`方法也调用遍历器接口，则可将`Generator`函数返回的`Iterator`对象，作为参数
```js
function* numbers () {
  yield 1;
  yield 2;
  return 3;
  yield 4;
}
[...numbers()];
// [ 1, 2 ];
let [ a, b ] = [...numbers()];
// a = 1; b = 2;
Array.from(numbers());
// [ 1, 2 ];
```

## 4. Generator.prototype.throw()
`Generator`函数返回的Iterator，都有一个`throw`方法，可在函数体外throw error,在Generator函数体内捕获
```js
function* g() {
  try {
    yield;
  } catch (error) {
    console.log(`内部捕获: ${error}`)
  }
}

let it = g();
it.next();

try {
  it.throw("a");
  it.throw("b");
}  catch (error) {
  console.log(`外部捕获：${error}`);
}

// 内部捕获: a
// 外部捕获：b
```

`Generator`函数内/外均未部署`try...catch`代码块时，程序将报错，直接中断执行
```js
function* gen () {
  yield console.log("hello");
  yield console.log("world");
}
let it = gen();
it.next();
it.throw();
// hello
// Uncaught undefined
```

`throw`方法抛出的错误要被内部捕获，前提必须至少执行过一次`next`方法
```js
function* gen() {
  try {
    yield 1;
  } catch (e) {
    console.log(e);
  }
}

let it = gen();
it.throw(1);
// uncaguth 1

// 上述代码，it.throw()执行时，next方法一次未执行，此时不会被内部捕获，而外层同样未捕获，程序终止报错
```
> Note: 可以理解为yeild先返回后面的值，再次运行时，throw会传入抛出的错误，再进行try...catch操作

`throw`方法捕获，也相同与调用了next方法，附带执行一条yield表达式
```js
function* gen () {
  try {
    yield console.log(1);
  } catch (e) {
    console.log(e);
  }
  yield console.log(2);
  yield console.log(3)
}

let it = gen();
it.next();    // 1
it.throw('Oops! Something wrong.'); // "Oops! Somethign wrong!"2 
it.next(); // 3
```

`throw` 与 `it.throw`方法无关，两者互补影响
```js
function* gen () {
  yield console.log(1);
  yield console.log(2);
}
let it = gen();
it.next();

try {
  throw new Error("Something wrong!");
} catch (e) {
  it.next();
}
// 1 2
```

>Generator 函数体外抛出的错误，可以在函数体内捕获；反过来，Generator 函数体内抛出的错误，也可以被函数体外的catch捕获。
```js
function* gen () {
  let x = yield 2;
  let y = x.toUpperCase();
  yield y;
}
let it = gen();
it.next();  // { value: 2, done: false }
try {
  it.next(42);
} catch (error) {
  console.log('Oops!', error);  
}
// Uncaught TypeError: x.toUpperCase is not a function
// undefined
```

>一旦 Generator 执行过程中抛出错误，且没有被内部捕获，就不会再执行下去了。如果此后还调用next方法，将返回一个value属性等于undefined、done属性等于true的对象，即 JavaScript 引擎认为这个 Generator 已经运行结束了。
```js
function* gen () {
  yield 1;
  console.log('throwing an exception');
  throw new Error('generator broke!');
  yield 2;
  yield 3;
}
function log(generator) {
  let v = undefined;
  console.log('Starting generator');
  try{
    v = generator.next();
    console.log('1st calls next method', v)
  } catch (e) {
    console.log('Catch error', v);
  }
  try{
    v = generator.next();   // 第二次的next并未执行
    console.log('2nd calls next method', v)
  } catch (e) {
    console.log('Catch error', v);
  }
  try{
    v = generator.next();
    console.log('3rd calls next method', v)
  } catch (e) {
    console.log('Catch error', v);
  }
  console.log('caller done');
}

log(gen());
// starting generator
// 第一次运行next方法 { value: 1, done: false }
// throwing an exception
// 捕捉错误 { value: 1, done: false }
// 第三次运行next方法 { value: undefined, done: true }
// caller done
```

## 5. Generator.prototype.return()
`Generator`函数返回的`Iterator`对象，还有`return`方法，返回给定的值，并终结遍历的`Generator`函数
```js
function* gen () {
  yield 1;
  yield 2;
  yield 3;
}
let it = gen();
it.next();
it.return('End');
it.next();
//  { value: 1, done: false }
//  { value: "End", done: true }
//  { value: undefined, done: true }
```
`return`不提供参数时，则返回值value值为undefined
```js
function* gen () {
  yield 1;
  yield 2;
  yield 3;
}
let it = gen();
it.next();
it.return();
// { value: 1, done: false }
// { value: undefined, done: true }
```
`Generator`函数内部有try...catch代码块，且正在执行try代码块，则`return`会立即进入`finally`代码块，执行完以后，整个函数才会结束
```js
function* numbers() {
  yield 1;
  try {
    yield 2;
    yield 3;
  } finally {
    yield 4;
    yield 5;
  }
  yield 6;
}
let it = gen();
it.next();
it.next();
it.return(7);
it.next();
it.next();
it.next();
// { value: 1, done: false };
// { value: 2, done: false };
// { value: 4, done: false };
// { value: 5, done: false };
// { value: 7, done: true };
// { value: undefined, done: true };
```

## 6. next, throw, return 三者关联

next()、throw()、return()这三个方法本质上是同一件事，可以放在一起理解。它们的作用都是让 Generator 函数恢复执行，并且使用不同的语句替换yield表达式。
- next()是将yield表达式替换成一个值
- throw()是将yield表达式替换成一个throw语句
- return()是将yield表达式替换成一个return语句

## 7. yield*表达式
如果在 Generator 函数内部，调用另一个 Generator 函数。需要在前者的函数体内部，自己手动完成遍历。
```js
function* gen () {
  yield "a";
  yield "b";
}
function* bar() {
  yield "x";
  for (let i of gen()) {
    console.log(i);
  };
  yield "y";
}

for (let v of bar()) {
  console.log(v);
}
// x a b y
```

`yield*`解决深层嵌套的问题
```js
function* gen () {
  yield "a";
  yield "b";
}

function* bar() {
  yield "x";
  yield* gen();
  yield "y";
}

// 等同于
function* bar() {
  yield 'x';
  yield 'a';
  yield 'b';
  yield 'y';
}

// 等同于
function* bar() {
  yield 'x';
  for (let v of foo()) {
    yield v;
  }
  yield 'y';
}

for (let v of bar()) {
  console.log(v);
}
// x a b y
```
`yield*`后面的Generator函数(无return语句时)，等同于在Generator函数内部，部署一个`for...of`循环
```js
function* gen() {
yield* ["a", "b", "C"];
}
for(let val of gen()) {
console.log(val);
}
// a b c
```
如果`yield`不加`*`,则会返回整个数组，加了之后，就调用了`for...of`循环进一步调用Iterator接口
```js
function* gen() {
  yield ["a", "b", "c"];
}
for(let val of gen()) {
  console.log(val);
}
// [ 'a', 'b', 'c' ]
```

实际是，任何数据结构只要有`Iterator`接口，就能被`yield*`遍历。正如`for...of`遍历循环
```js
function* gen() {
  yield "helllo";
  yield* "hello";
}
for (let val of gen()) {
  console.log(val);
}
// helllo h e l l o
```

如果被代理的 `Generator` 函数有`return`语句，那么就可以向代理它的 `Generator` 函数返回数据。
> 例一：
```js
function* foo() {
  yield 2;
  yield 3;
  return "foo";   
  // 被代理使用时，return会直接返回值，不再在next调用返回
}

function* bar() {
  yield 1;
  let val = yield* foo();
  console.log('val: ', val);
  yield 4;
}
let it = bar();
it.next();  // {value: 1, done: false}
it.next();  // {value: 2, done: false}
it.next();  // {value: 3, done: false}
it.next();  // {value: 4, done: false}
it.next();  // {value: undefined, done: true}
```
>例二：被代理的generator函数有`return`返回值
```js
function* foo() {
  yield "a";
  yield "b";
  return "The result is"
}
function* bar() {
  let result = yield* foo();
  console.log(result);
}

[...bar()];
// The result is
// [ "a", "b" ]
```
`yield*`命令可以很方便地取出嵌套数组的所有成员，或平铺嵌套数组。
```js
function* iterTree(tree) {
  if (Array.isArray(tree)) {
    for (let val of tree) {
      yield* iterTree(val);
    } 
  } else {
    yield tree;
  }
}
let tree = ["a", [ "b", [ "c", "d"]], "e"];
[...iterTree(tree)];
// ["a", "b", "c", "d", "e"]
```

`yield*`遍历完全二叉树
```js
class Tree {
  constructor (left, label, right) {
    this.left = left;
    this.label = label;
    this.right = right;
  }
}
// 中序遍历函数
function* inorder(tree) {
  if (tree) {
    yield* inorder(tree.left);
    yield tree.label;
    yield* inorder(tree.right);
  }
}

// 生成二叉树
function make(array) {
  return array.length == 1 ?
    new Tree(null, array[0], null) :
    new Tree(make(array[0]), array[1], make(array[2]));
}
let tree = make([[['a'], 'b', ['c']], 'd', [['e'], 'f', ['g']]]);

// 遍历二叉树
let result = [];
for (let node of inorder(tree)) {
  result.push(node);
}
// ["a", "b", "c", "d", "e", "f", "g"]
```

## 8. Generator as property
> 对象属性简写方式
```js
// shortcut
let obj = {
  * gen() {
    /*...*/
  }
}

// complete version
let obj = {
  gen: function* () {
    /*...*/
  }
}
```

## 9. Generator this
`Generator`函数返回的遍历器是`Generator`的实例
```js
function* gen() {};
gen.prototype.say = function () {
  return "Bonjour";
}
let g = gen()

g instanceof gen;   
// true
g.say();
// "Bonjour"
```
如果把`gen`当作普通的构造函数，并不会生效，因为`gen`返回的总是遍历器对象，而不是`this`对象。
> this指向的是window全局对象
```js
function* gen () {
  console.log(this);
  this.a = 11;
}
let obj = gen();   
// 没有return默认为undefined
obj.next();
// this指向window
// {value: undefined, done: true}
// 为this添加的a，直接加到了window对象全局属性
obj.a 
// undifined

```
`Generator` 函数也不能跟`new`命令一起用，会报错。
```js
function* F() {
  yield this.x = 2;
  yield this.y = 3;
}
new F();
// Uncaught TypeError: F is not a constructor
```

- 变通方法一：
> 首先，生成一个空对象，使用call方法绑定 Generator 函数内部的this。这样，构造函数调用以后，这个空对象就是 Generator 函数的实例对象了。
```js
function* F() {
  this.a = 1;
  yield this.b = 2;
  yield this.c = 3;
}
let obj = {};
let f = F.call(obj);
// 这样添加的a, b, c属性都会在obj对象上
f.next();
// {value: 2, done: false}
f.next();
// {value: 3, done: false}
f.next();
// {value: undefined, done: true}
obj.a;
// 1
obj.b;
// 2
obj.c;
// 3
```

- 变通方法二：
一个办法就是将`obj`换成`F.prototype`。
```js
function* F() {
  this.a = 1;
  yield this.b = 2;
  yield this.c = 3;
}
let f = F.call(F.prototype);
f.next();  // Object {value: 2, done: false}
f.next();  // Object {value: 3, done: false}
f.next();  // Object {value: undefined, done: true}

f.a // 1
f.b // 2
f.c // 3
F.prototype;
// Generator {a: 1, b: 2, c: 3}
Object.getPrototypeOf(f);
// Generator {a: 1, b: 2, c: 3}
```
再将`F`改成构造函数，就可以对它执行`new`命令了
```js
function* gen() {
  this.a = 1;
  yield this.b = 2;
  yield this.c = 3;
}
class F {
  constructor() {
    return gen.call(gen.prototype)
  }
};
let f = new F();
f.next();  // Object {value: 2, done: false}
f.next();  // Object {value: 3, done: false}
f.next();  // Object {value: undefined, done: true}

f.a // 1
f.b // 2
f.c // 3
f instanceof gen;
// true
```

## 10. explanation
### 10.1 Generator与状态机 state machine
```js
// 对比
// ES5 实现方式
let ticking = true;
let clock = function () {
  ticking ? console.log("Tick!") : console.log("Tock!");
  ticking = !ticking;
}
while(true) {
  clock();
}

// ES6
function* clock() {
  while(true) {
    console.log("Tick!");
    yield;
    console.log("Tock!");
    yield;
  }
}
let c =  clock();
c.next();
c.next();
c.next();
// Tick!  {value: undefined, done: false}
// Tock!  {value: undefined, done: false}
// Tick!  {value: undefined, done: false}
```
>Note: 上面的 Generator 实现与 ES5 实现对比，可以看到少了用来保存状态的外部变量ticking，这样就更简洁，更安全（状态不会被非法篡改）、更符合函数式编程的思想，在写法上也更优雅。Generator 之所以可以不用外部变量保存状态，是因为它本身就包含了一个状态信息，即目前是否处于暂停态。

### 10.2 Generator与协程 coroutine
协程（coroutine）是一种程序运行的方式，可以理解成“协作的线程”或“协作的函数”。协程既可以用单线程实现，也可以用多线程实现。前者是一种特殊的子例程，后者是一种特殊的线程。

- 后进先出
- semi-coroutine半协程

### 11. application运用
- 1. 异步操作的同步化表达
```js
function* loadUI() {
  showLoadingScreen();
  yield loadUIDataAsynchronously();
  hideLoadingScree();
}

function showLoadingScreen() {
  console.log("showLoadingScreen");
}
function loadUIDataAsynchronously() {
  console.log("loadUIDataAsynchronously");
  let result = undefined;
  setTimeout(()=>Promise.resolve("loadUIDataAsynchronously").then(data => result = data ),1000);
  return "result " + result;
}
function hideLoadingScree() {
  console.log("hideLoadingScree");
}
let loader = loadUI();
loader.next();
loader.next();

// Generator函数逐行读取文本文件
function* numbers() {
  let file = new FileReader("./numbers.txt");
  try {
    while (!file.eof) {
      yield parseInt(file.readLine(), 10);
    }
  } finally {
    file.close();
  }
}
let n = numbers();
n.next();
```

- 2. 控制流管理
- 3. 部署Iterator接口
- 4. 作为数据结构

## 12. Generator函数的异步应用
### 12.1 Traditional methods 传统方法
ES6 诞生以前，异步编程的方法，大概有下面四种。
 - 回调函数
 - 事件监听
 - 发布/订阅
 - Promise 对象
Generator 函数将 JavaScript 异步编程带入了一个全新的阶段。

### 12.2 Generator函数的数据交换 和 错误处理

### 12.3 异步任务的封装
```js
const axios = require('axios');
let url ="https://api.github.com/users/github";
function* gen() {
  let result = yield axios(url);
  console.log(result);
}

// console.log(axios(url).then(data => console.log(data)));
let g = gen();
g.next().value.then(data => data.data).then(data => g.next(data))
```

### 12.4 Thunk函数
Thunk 函数是自动执行 Generator 函数的一种方法。

- 参数的求值策略

- Thunk函数的含义

- js语言的Thunk函数

- `Thunkify`模块

- `Generator`函数的流程管理

- Thunk 函数的自动流程管理

### 12.5 co模块
